### DISCUSSION TITLE
upload limit and hot reload of config

### DISCUSSION THREAD

**Discussion by tessus** [orhun/rustypaste#93] [General]
There are 2 things I came across that seemed a bit strange to me. 

1. The content file size limit is only evaluated AFTER a file has been uploaded.
2. Hot reload of the configuration does not seem to work.

I set the value to 50MB, because sometimes log files even in compressed form can be quite large.
It happend that I had a 710MB compressed log file and forgot that I had the limit in place.
It uploaded the entire 710MB only to tell me that it exceeded the content length limit.
This is just an extreme case and probably won't happen that often, but I think the limit should be evaluated before any transfer happens.
Afaik the content length is part of the header for POST requests so it should be fairly easy to test for length compliance.

As for 2, I am not really using the feature, but I added a config value and changed another and neither was picked up after the "hot reload" refresh time lapsed. I looked at the code, but I didn't fully understand how the original config struct was actually changed/replaced. It only changes a cloned version unless I missed something. I would have to do more tests, but maybe somebody is using the feature and has some additional input.

**Comment by tessus** [comment]
I did more testing and the middleware reports the `Upload rejected due to exceeded limit.` warning in the log file right away, but the file is still uploaded by the client (curl, rpaste).
Only after the file has been uploaded, the error message appears on the client side.

Thus something is not ok with the middleware. Unfortunately I don't understand this part of the code.

**Reply by tessus** [reply]
First I tested with smaller files, less than 10000 bytes. When I use a limit of 50MB and upload a file that is 70MB:

- curl hangs
- rpaste uploads the file completely and receives an error `upload limit exceeded (status code: 413)`

Something is definitely off.

**Reply by orhun** [reply]
Unfortunately I can't reproduce this. I set the `max_content_length` to "50mb" and tried to upload 70mb file (`truncate -s 70MB somefile`). The file isn't uploaded and I got the "upload limit exceeded" message.

**Reply by tessus** [reply]
Hmm, I have run into this issue on 2 separate instances. Did you use an auth_token? I will do a few more tests later this afternoon without an auth_token set.
This is the only thing I can think of that might be different in our envs.
Or maybe the reverse proxy config?

Ah, I will run some more tests trying to narrow it down.

**Reply by orhun** [reply]
Ah, no I didn't use a auth token in my tests.

> Ah, I will run some more tests trying to narrow it down.

Alright, sounds good!

**Reply by tessus** [reply]
I have now tested with and without auth_tokens. As well as with and without reverse proxy.
Of course all permutations of the above.

It is reproducible all the time:

- curl hangs
- rpaste uploads the entire file and gets the error message (although the server gets the error message right away)

**Reply by orhun** [reply]
Hmm, let's add a new test fixture to test this on GitHub Actions as well? How does that sound?

**Reply by tessus** [reply]
I can look into this. First I want to get the auth_tokens PR done.

**Reply by orhun** [reply]
Alright!

**Reply by tessus** [reply]
There are fixtures that test the content limiter. Strangely enough, when I run these fixtures locally on my macOS, they do not fail. I even tried to add a huge file to that particular test, but curl doesn't hang. (Thinking out loud: Is is because client and server are both on macOS - still, this doesn't make sense.)

My test was macOS client with server on Linux (w/ and w/o reverse proxy, w/ and w/o auth). I am going to test with Linux boxes and see, if I can see the same issue.

**Reply by orhun** [reply]
I see, I would still say let's add another fixture to testing huge files on GitHub Actions.

**Reply by tessus** [reply]
The problem is how to detect whether the error message shows up before the file is uploaded or after? And how can I make sure that the file was not in fact uploaded at all.

I have to think about that...

**Comment by orhun** [comment]
> The content file size limit is only evaluated AFTER a file has been uploaded.

This was supposed to be fixed in #53 

> Hot reload of the configuration does not seem to work.

For some of the values such as `upload_path`, `timeout` and `max_content_length` cannot be hot-reloaded (AKA configured on the fly) because they are supposed to be configured at the server start.

> As for 2, I am not really using the feature, but I added a config value and changed another and neither was picked up after the "hot reload" refresh time lapsed. I looked at the code, but I didn't fully understand how the original config struct was actually changed/replaced. It only changes a cloned version unless I missed something. I would have to do more tests, but maybe somebody is using the feature and has some additional input.

Can you set `RUST_LOG` environment variable to `trace` and see if there are any logs about the hot-reloader? (e.g. "watcher dropped" etc.)

**Reply by tessus** [reply]
> This was supposed to be fixed in https://github.com/orhun/rustypaste/issues/53

Yes, this is what I gathered from the PR notes and the comments in the code. I saw that stackoveflow post regarding having to drain the payload. So I thought that there is still some edge case where actix might be off. I am not sure why the dev doesn't fix this as a high priority bug. Everything else is not important at that point.
I have only dabbled in one other Rust project (vaultwarden) and there rocket was used. So actix is also new to me.

> For some of the values such as upload_path, timeout and max_content_length cannot be hot-reloaded (AKA configured on the fly) because they are supposed to be configured at the server start.

How do people know which those are?

> Can you set RUST_LOG environment variable to trace and see if there are any logs about the hot-reloader? (e.g. "watcher dropped" etc.)

I can do that. I should be able to get this done in the next few days.

**Reply by orhun** [reply]
> I am not sure why the dev doesn't fix this as a high priority bug. Everything else is not important at that point.

Yeah, there might be some actix sorcery there which might be causing this or not. Frankly, the whole middleware implementation feels a bit hacky to me but yeah...

> How do people know which those are?

Those are the values which are being used before the hot-reloader library is initialized. It is not mentioned anywhere but I thought it would be clear that e.g. you can't change the server address after the server has started :D Either way, feel free to take a look into it and make a note in README.md or even better make them hot-reloadable. Don't know if it is possible. 

> I can do that. I should be able to get this done in the next few days.

Great, thanks! There were some regressions about that in the past (a13c0f123a3d0d27ed5c6b879f475bc1e25f84c9) so I was just wondering if it is related to that in any way.

**Reply by tessus** [reply]
> Frankly, the whole middleware implementation feels a bit hacky to me but yeah...

I don't understand that part yet. It's a bit too much new Rust for me. ;-) Working on it.

> you can't change the server address after the server has started :D

I know, this makes total sense and I am not saying to change this. I just wanted to know, where in the code I can see the list of values that cannot be changed. Because I did not see any exclusions in the reload code. But as I mentioned, I don't fully understand that part anyway, because only the copy is changed, not the original. Or maybe it is a reference. I have to check again. I only quickly browsed over it.

**Reply by orhun** [reply]
So here is the part you can look at:

https://github.com/orhun/rustypaste/blob/f09c467f6e25d446adb2a7410ebc0859e7fd8372/src/main.rs#L51-L96

Since `server_config` is cloned, it is not being updated by the hotwatcher thus the hot-reloaded values do not take effect.

**Reply by tessus** [reply]
right, and the server config also holds the auth_tokens. There are 3 options in server_config that would make sense to hot reload:

- expose_version
- auth_tokens
- max_content_length

Everthing else should not be changed. Need to dive into the code. Maybe I can make that happen somehow.

**Reply by tessus** [reply]
> Since `server_config` is cloned, it is not being updated by the hotwatcher thus the hot-reloaded values do not take effect.

I was looking into the code, but it still makes no sense to me. Yes, `server_config` is cloned, but the hotwatcher uses these 2 vars:

```rust
let config = Data::new(RwLock::new(config)); 
let cloned_config = Data::clone(&config); 
```

And they reference the entire config struct. So I am not sure, why `auth_tokens` or `expose_version` are not hot reloaded.

**Reply by tessus** [reply]
Actually they are hot reloaded. I am not sure why that didn't happen in my previous tests. The only difference was the refresh interval. Previously I had set it to `1h` because I don't think it is necessary to check/reload every second. But even after 3 hours waiting, the changes were not picked up. I will run some more tests with an interval of 1h.

But now I was testing with `1s` and changes to `auth_tokens` and `expose_version` were picked up and hot reloaded.

**Reply by tessus** [reply]
I did some more tests with an interval of 1h. But it seems that the config parameters I had in mind are in fact hot reloaded. I don't know why it didn't work when I started this discussion.
Maybe the quantum signature was slightly distored due to quarks being unhappy. It's a mystery to me.

I guess this part of the discussion is resolved.

**Reply by orhun** [reply]
Sounds like everything is working as expected! (fingers crossed)
